### ISSUE TITLE
<b-table>, when using a provider function, can't have Object filters

### ISSUE THREAD

**Issue by niels** [bootstrap-vue/bootstrap-vue#3428]
### Describe the bug

When its `filter` prop is an `Object`, `<b-table>` does not call its provider function when the prop's contents change.

### Steps to reproduce the bug

```vue
<template>
  <div>
    <b-table
      :items="loadItems"
      :filter="filter"
    />

    <b-input v-model="filter.foo" />
  </div>
</template>

<script>
export default {
  data() {
    return {
      filter: { foo: null },
    };
  },

  watch: {
    filter: function(newVal, oldVal) {
      console.log('Filter reactivity', this.filter, oldVal, newVal);
    },
    'filter.foo': function(newVal, oldVal) {
      console.log('Filter key reactivity', this.filter.foo, oldVal, newVal);

      // Trigger reactivity even on the root this.filter to guarantee that all watchers fire.
      this.filter = { ...this.filter };
    },
  },

  methods: {
    loadItems() {
      console.log('Loading items with filter', this.filter);
      return [{ col: '1' }];
    },
  },
};
</script>
```

### Actual behavior

1. When typing anything into the input field,
2. The `watch`ers fire in the above component; and
3. The `filter` prop within `<b-table>` is update correctly; but
4. `loadItems` is not called.

Console output:

```
Loading items with filter …
Filter key reactivity …
Filter reactivity …
```

### Expected behavior

Whenever the `filter` prop of `<b-table>` changes, and `items` is a function, that function should be called.

I have not yet contributed to this project and may thus misunderstand the expected behavior. But I believe that https://github.com/bootstrap-vue/bootstrap-vue/blob/ae2040b2dfa7eba75d0e6468bf7321b5ac6e8ec5/src/components/table/helpers/mixin-filtering.js#L56 indicates that the filter prop should work fine when given an Object.

### Versions

**Libraries:**

- BootstrapVue: 2.0.0-rc.21
- Bootstrap: 4.3.1
- Vue: 2.6.10

**Environment:**

- Device: Desktop
- OS: Arch Linux
- Browser: Firefox
- Version: 67

**Comment by tmorehouse** [comment]
We may need to implement deep inspection of the `filter` prop (via our `looseEqual` utility)

**Comment by tmorehouse** [comment]
Found the issue.  In your case, the filter object is a Vue reactive object, with Vue observers attached (which makes the properties of the object as getters/setters.  

In PR #3429, we have now wrapped the generation of hte localFilter computed property with a `cloneDeep` util, which converts it back to a plain object. As well, we needed to add in a check to see if `localFilter` functions were being used or if the provider would be doing the filtering (as it was always returning an empty string when filter was an object and items was a provider).

PR #3429 fixes your issue, and will be available in the next release (v2.0.0-rc.23)
